### PR TITLE
Compute confidence intervals in DistributedTreeDriver

### PR DESCRIPTION
**Pull request by masterleinad**
Based on #78, this pull requests improves `DistributedTreeDriver` further. The idea is as follows:
- We first run the test a couple of time to have an estimate for mean and variance of the runtime for the individual sections (`construction`, `knn`, `radius`). (We take the maximum over all MPI processes in each iteration).
- These statistics are used to estimate for each section the number of total iterations needed to get a confidence interval of specified width.
- We run the maximum of these estimates number iterations (additionally) and output mean, variance and a new confidence interval at the end.

All this is based on `CppCon 2015: Bryce Adelstein-Lelbach “Benchmarking C++ Code` and `Boost.Accumulators`.

Sample output:
```
$ mpiexec -np 2 ./ArborX_DistributedTree.exe
ArborX version: 0.9 (dev)
ArborX hash   : 829b702

Running with arguments:
perform knn search      : true
perform radius search   : true
#points/MPI process     : 50000
#queries/MPI process    : 20000
size of shift           : 1
dimension               : 3


Sample lap 0:
contruction done
knn done
radius done

[...]

Sample lap 9:
contruction done
knn done
radius done

estimated 11 iterations

Total lap 10:
contruction done
knn done
radius done

[...]

Total lap 20:
contruction done
knn done
radius done
=========================================================================

TimeMonitor results over 2 processors

Timer Name   | mean         | variance     | confidence interval         
-------------------------------------------------------------------------
construction | 5.052878e-02 | 2.047170e-05 | [4.841836e-02, 5.263920e-02]
knn          | 1.172640e+00 | 2.565172e-02 | [1.097935e+00, 1.247345e+00]
radius       | 6.593428e-01 | 6.237161e-03 | [6.225057e-01, 6.961799e-01]
=========================================================================
```

### CODE REVIEW

File: examples/distributed_tree_driver/distributed_tree_driver.cpp [hunk 1]
```diff
@@ -333,27 +383,134 @@ int main_(std::vector<std::string> const &args)
...
+    std::cout << "\nRunning with arguments:\n"
+              << "perform knn search      : " << arguments.perform_knn_search
+              << '\n'
+              << "perform radius search   : " << arguments.perform_radius_search
+              << '\n'
+              << "#points/MPI process     : " << arguments.n_values << '\n'
+              << "#queries/MPI process    : " << arguments.n_queries << '\n'
+              << "size of shift           : " << arguments.shift << '\n'
+              << "dimension               : " << arguments.partition_dim << '\n'
+              << '\n';
```
Comment (dalg24) [inline examples/distributed_tree_driver/distributed_tree_driver.cpp]: Please consider adding a "print" member function to `Arguments` or a "put to" operator that write that info into a stream.

**Comment by masterleinad** [comment]
I have no idea why compiling with `nvcc_wrapper` fails...

File: examples/distributed_tree_driver/distributed_tree_driver.cpp [hunk 2]
```diff
@@ -27,68 +32,130 @@
...
+          boost::math::quantile(complement(dist, (1 - confidence) / 2));
+
+      // We need to scale the sample standard deviation to get an unbiased
+      // estimator of the population standard deviation.
+      double const current_stddev = std::sqrt(
+          ba::variance(_statistics) * (n_measurements / (n_measurements - 1.)));
+      double const current_mean = ba::mean(_statistics);
+      auto const tmp =
+          z * current_stddev / (relative_error_margin * current_mean / 2.);
+      return static_cast<int>(std::ceil(tmp * tmp));
```
Comment (masterleinad) [inline examples/distributed_tree_driver/distributed_tree_driver.cpp]: So far I couldn't find parameters for `boost::math::students_t::find_degrees_of_freedom` such that the returned value would match the one computed here.
```
boost::math::students_t::find_degrees_of_freedom(relative_error_margin*current_mean, confidence/2, .9999999, current_stddev)
```
is just 1.3 times larger though.

**Comment by masterleinad** [comment]
> I have no idea why compiling with `nvcc_wrapper` fails...

Just having a 
```
 boost::accumulators::accumulator_set<double, ba::stats<ba::tag::count>>;
```
member variable gives
```
/opt/boost/include/boost/fusion/iterator/mpl/convert_iterator.hpp(57): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/iterator/mpl/convert_iterator.hpp(57): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/view/filter_view/filter_view_iterator.hpp(60): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/view/filter_view/filter_view_iterator.hpp(60): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/algorithm/query/detail/find_if.hpp(208): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/algorithm/query/detail/find_if.hpp(208): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/algorithm/query/detail/find_if.hpp(208): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/iterator/next.hpp(61): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/algorithm/query/detail/find_if.hpp(208): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/algorithm/query/detail/find_if.hpp(208): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/algorithm/query/detail/find_if.hpp(208): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/algorithm/iteration/detail/for_each.hpp(48): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/algorithm/iteration/detail/for_each.hpp(36): error: identifier "" is undefined in device code
/opt/boost/include/boost/fusion/algorithm/iteration/detail/for_each.hpp(48): error: identifier "" is undefined in device code
```
when compiling with `nvcc_wrapper`.

**Comment by aprokop** [comment]
Anyone opposed to closing this? It was a nice attempt, but too complicated with little benefit in practice.

**Comment by masterleinad** [comment]
Yes, I think we won't do that.

**Pull request closed without merging**